### ISSUE TITLE
Legacy X11 tray-icons don't get keyboard focus

### ISSUE THREAD

**Issue by jwrdegoede** [ubuntu/gnome-shell-extension-appindicator#301]
When running a native-wayland app and then right-clicking on a legacy X11 tray-icon, keyboard navigation of the popup-menu does not work.

I'm pretty sure this is happening because this uses a keyboard-grab, but that only grabs the kbd events at the xserver level and with the current focussed app being a wayland app the events are never send to the xserver.

If the currently running/focussed app is a X11 app then the problem does not happen.

Perhaps we can make the popup window itself an input/output window and have it get the keyboard focus ?

We should probably ask @ofourdan to help here, since he knows the xwayland keyboard grabbing code the best.

This assumes that we deem fixing this worth the trouble (we are talking about legacy stuff here after all).

**Comment by ofourdan** [comment]
Does it helps to allow all X11 clients to issue keyboard grabs?

```
gsettings set org.gnome.mutter.wayland xwayland-allow-grabs true
gsettings set org.gnome.mutter.wayland xwayland-grab-access-rules "['*']"
```

**Comment by jwrdegoede** [comment]
Yes that helps and things still work when I replace the `*` with the wm-class of the app from which the trayicon is.

So I guess everything is working as intended, IOW this is not a bug ?

Specifically I'm wondering if there is anything we can do to allow grabs automatically in this case, but from wayland/X11's pov the popup is just as any other window, so I'm pretty sure the answer is no.

**Comment by ofourdan** [comment]
I would prefer to avoid the use of hacks for this (the whole thing is already a hack, including the complete utter nonsense that is the X11 tray icon), and I am not sure this is actually necessary.

An X11 client wishing to be granted Xwayland grabs by gnome-shell/mutter can send a `ClientMessage` to the root window with:
    
     - message_type set to "_XWAYLAND_MAY_GRAB_KEYBOARD"
     - window set to the xid of the window on which the grab is to be issued
     - data.l[0] to a non-zero value

See https://gitlab.gnome.org/GNOME/mutter/-/commit/5f132f397

**Comment by 3v1n0** [comment]
@ofourdan that could be probably implemented as part of the shell's tray manager?

**Comment by ofourdan** [comment]
Maybe, but I don't really know that code.

**Comment by 3v1n0** [comment]
I've been looking a bit, and probably we could make the event to be sent when the `ShellTrayIcon` is constructed since there we've the ability to get the XID of the window.

I suppose we may also try to get it from the extension, iterating all the windows whose `_NET_WM_PID` is matching the one of the icon we get, but it seems a bit too much to me.
